## Creating My Personalized Sample

## Import pandas 

In [1]:
import pandas as pd

# Load master dataset
df_wdi_full = pd.read_csv("wdi_dataset.csv")

# Define your student ID
student_id = "nf1028810"  # Replace with your own ID
SAMPLE_FRACTION = 0.15
SEED_DIGITS = 3 

# Derive random seed
seed = int(str(student_id)[-SEED_DIGITS:])

# Create your sample
df_sample = df_wdi_full.sample(frac=SAMPLE_FRACTION, random_state=seed).reset_index(drop=True)

# Save the sample
filename = f"student_sample_{student_id}.csv"
df_sample.to_csv(filename, index=False)

print(f"Your sample ({len(df_sample)} rows) has been saved as {filename}")
print(f"Seed used: {seed}")

Your sample (1232 rows) has been saved as student_sample_nf1028810.csv
Seed used: 810


**Interpretation:**

The code defines a standardized process for creating a personalized, reproducible sample for analysis. It uses the student ID to derive a random seed 810 from nf1028810 and then samples 15\% of the master dataset (df_wdi_full). This ensures that the generated sample is unique to the student but can be recreated consistently, providing a fixed dataset for all subsequent tasks.

## TASK 1 – Dataset Overview

In [2]:
#1	Display number of rows and columns

# Load the saved sample file
df_sampled_data = pd.read_csv(filename)

# Create a variable for rows and column to display differently
num_rows, num_columns = df_sampled_data.shape
print(f"The sampled dataset contains {num_rows} rows and {num_columns} columns.")


The sampled dataset contains 1232 rows and 10 columns.


**Interpretation:**

I got the number of rows and columns using .shape and created different variables to show the result separatly.
The dataset contains 1232 observations and 10 features.


In [6]:
#1.2 	List column names and data types.

print("\nColumn Names and Data Types:")
print(df_sampled_data.info())


Column Names and Data Types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1232 entries, 0 to 1231
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            1232 non-null   object 
 1   Year               1232 non-null   int64  
 2   LifeExpectancy     1232 non-null   float64
 3   GDPperCapita       1174 non-null   float64
 4   Population         1232 non-null   float64
 5   SchoolEnrollment   876 non-null    float64
 6   InternetUsers      900 non-null    float64
 7   HealthExpenditure  742 non-null    float64
 8   LiteracyRate       296 non-null    float64
 9   GDPGrowth          1152 non-null   float64
dtypes: float64(8), int64(1), object(1)
memory usage: 96.4+ KB
None


**Interpretations:**

I displayed the column names and data types using df.info(), because The df.info() output also reveals a significant amount of missing data in columns like SchoolEnrollment, InternetUsers, and notably LiteracyRate.

In [7]:
#1.3 	Show the range of years and number of unique countries.

#3.1 Calculate the minimum and maximum year values
min_year = df_sampled_data['Year'].min()
max_year = df_sampled_data['Year'].max()

#3.2 Calculate the number of unique countries
num_unique_countries = df_sampled_data['Country'].nunique()

print(f"Range of Years: {min_year} - {max_year}")
print(f"Number of Unique Countries: {num_unique_countries}")

Range of Years: 1990 - 2020
Number of Unique Countries: 261


**Interpretation:**

1.  Range of Years:
    - I used **(.min())** to find the smallest and largest **(.max())** values in that column.
    - This tells us the time span covered by the sampled data. Since the result is **1990 - 2020** which is 31 years, providing a good longitudinal view.

2.  Number of Unique Countries:
    - I count the number of unique values in the Country column using **.nunique()**.
    - The sample covers 261 unique countries, which is a substantial representation of the global dataset.

## TASK 2 – Filtering Data

In [8]:
#2.1 	Filter data for years ≥ 2000 and count rows.

# Filter the DataFrame for rows where 'Year' is 2000 or later
count_2000_onwards = df_sampled_data[df_sampled_data['Year'] >= 2000]

print(f"Number of rows with Year >= 2000: {len(count_2000_onwards)}")

Number of rows with Year >= 2000: 836


**Interpretation:**

This filter isolates data points from the start of the 21st century onwards. Out of 1232 total observations, 814 rows are from 2000 or later, showing that roughly two-thirds of the sample is from the past two decades.


In [9]:
#2.2 	Find observations where LifeExpectancy > 75 and GDPperCapita > 30,000.

# Checking for rows where LifeExpectancy > 75 and GDPperCapita > 30,000
df_life_gdp = df_sampled_data[
            (df_sampled_data['LifeExpectancy'] > 75) & 
            (df_sampled_data['GDPperCapita'] > 30000)
        ]
df_life_gdp



,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth
2,United Kingdom,2017,81.256098,40629.234035,65966000.0,100.781052,90.4246,3907.822266,NaN,2.656505
6,United Kingdom,2020,80.331756,40404.806224,66744000.0,102.223953,94.8182,4816.152832,NaN,-10.296919
33,Japan,2018,84.210976,39751.133098,126811000.0,102.099907,88.7464,4265.779297,NaN,0.643391
36,Israel,2010,81.602439,31439.088703,7623600.0,97.658630,67.5000,2237.907959,NaN,5.534340
41,San Marino,2008,82.894000,75902.132052,31662.0,94.248047,54.5200,4378.248535,NaN,-0.500939
...,...,...,...,...,...,...,...,...,...,...
1205,Monaco,1998,81.605000,90947.205122,32267.0,132.251526,NaN,NaN,NaN,3.503266
1212,"Macao SAR, China",2018,83.678049,83761.817180,658900.0,101.238255,83.7941,NaN,NaN,6.405279
1216,Liechtenstein,2018,83.041463,174988.775083,38246.0,106.329788,NaN,NaN,NaN,NaN
1228,Cayman Islands,2011,76.562000,75280.960765,55606.0,103.911343,69.4659,NaN,NaN,1.169062


**Interpretation:**

There are 149 observations where a country-year entry has a life expectancy over 75 years and a GDP per capita over $30,000. These rows represent highly developed and healthy societies in the sample, which are typically found in Europe, North America, and parts of Asia and the Middle East.

In [10]:
#2.3 	List the first 10 countries meeting those conditions.

df_life_gdp.head(10)

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth
2,United Kingdom,2017,81.256098,40629.234035,65966000.0,100.781052,90.4246,3907.822266,NaN,2.656505
6,United Kingdom,2020,80.331756,40404.806224,66744000.0,102.223953,94.8182,4816.152832,NaN,-10.296919
33,Japan,2018,84.210976,39751.133098,126811000.0,102.099907,88.7464,4265.779297,NaN,0.643391
36,Israel,2010,81.602439,31439.088703,7623600.0,97.658630,67.5000,2237.907959,NaN,5.534340
41,San Marino,2008,82.894000,75902.132052,31662.0,94.248047,54.5200,4378.248535,NaN,-0.500939
56,Qatar,2007,77.741000,65954.341953,1208595.0,98.263962,37.0000,1301.496826,93.080002,17.985657
58,Faroe Islands,2017,82.143902,59387.353086,50180.0,NaN,97.5820,NaN,NaN,3.512014
64,Sweden,2017,82.409756,53300.858728,10057698.0,122.896057,93.0063,5837.770996,NaN,1.825156
79,United Kingdom,2015,80.956098,44983.885523,65088000.0,100.204041,92.0003,4440.738281,NaN,2.222888
85,Canada,2010,81.322195,47560.666601,34005902.0,99.134369,80.3000,5062.482910,NaN,3.090806


**Interpretation:**

After filtering the data i used .head() to print the first 10 rows.

## TASK 3 – Grouping and Aggregation

In [11]:
#3.1 	Compute average life expectancy and average GDP per capita by country.

# Average Life Expectancy and GDPperCapita
df_avg_life = df_sampled_data.groupby('Country')[['LifeExpectancy', 'GDPperCapita']].mean()

print(f"The average GDPperCapita is: {df_avg_life}")

The average GDPperCapita is:                              LifeExpectancy  GDPperCapita
Country                                                  
Afghanistan                       53.374167    488.761738
Africa Eastern and Southern       57.079252   1114.720293
Africa Western and Central        55.958497   1777.474838
Albania                           76.883667   2839.779346
Algeria                           69.510667   2310.100708
...                                     ...           ...
West Bank and Gaza                72.552167   2658.849791
World                             70.291431   8592.940013
Yemen, Rep.                       63.835167    763.528644
Zambia                            51.609000    574.746795
Zimbabwe                          53.177167    940.278046

[261 rows x 2 columns]


**Interpretation:**

This aggregation provides a static, long-term average metric for each country, allowing for a high-level comparison of development and health outcomes across the entire sample period. For example, Albania has a high average life expectancy (76.88 years) relative to its modest average GDP per capita ($2,839.78), while regional aggregates like Africa Western and Central show lower averages in both metrics.

In [12]:
#3.2 	Display the top 5 countries for each measure.

#3.2.1 Top 5 countries by Life Expectancy
top_life = df_sampled_data.sort_values("LifeExpectancy", ascending=False).head(5)


top_life



,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth
1151,San Marino,2018,85.09500,47950.688056,34522.0,95.126050,NaN,3969.364746,NaN,1.493905
635,Monaco,2011,85.01000,179363.984222,33946.0,NaN,80.3000,8302.794922,NaN,7.022357
1049,"Hong Kong SAR, China",2017,84.69561,46160.429791,7393200.0,104.807910,89.4159,NaN,NaN,3.796114
688,Andorra,2016,84.48900,40129.819201,72181.0,91.041577,89.7000,2769.152832,NaN,3.709678
542,San Marino,2014,84.39400,51260.493835,32655.0,NaN,NaN,4120.026855,NaN,-0.702371


**Interpretation:**

The country-year observation with the highest life expectancy is San Marino in 2018 (85.095 years). This list highlights some of the world's longest-living populations.

In [13]:

#3.2.2 Top 5 countries by GDP per capita
top_gdp = df_sampled_data.sort_values("GDPperCapita", ascending=False).head(5)

top_gdp

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth
635,Monaco,2011,85.010000,179363.984222,33946.0,NaN,80.3000,8302.794922,NaN,7.022357
793,Liechtenstein,2014,82.073171,178735.153021,37248.0,102.660988,95.2100,NaN,NaN,NaN
1216,Liechtenstein,2018,83.041463,174988.775083,38246.0,106.329788,NaN,NaN,NaN,NaN
303,Liechtenstein,2013,82.260976,172823.607795,36984.0,NaN,93.8000,NaN,NaN,NaN
1063,Liechtenstein,2019,84.160976,166907.839305,38563.0,103.830009,93.1461,NaN,NaN,NaN


 **Interpretation:**
 
 Monaco in 2011 has the highest GDP per capita ($179,364), followed by Liechtenstein. This list highlight some of the world's wealthiest, often represented by small, high-income territories

In [14]:
#3.3 	Compute the average GDP growth rate per decade

# Show only data from 1990–2019
df_sampled_data = df_sampled_data[(df_sampled_data["Year"] >= 1990) & 
                                  (df_sampled_data["Year"] <= 2019) & (df_sampled_data["Year"] != 2020)]

# Create Decade column for the years
df_sampled_data["Decade"] = (df_sampled_data["Year"] // 10) * 10

# Compute average GDP growth per decade
avg_growth = df_sampled_data.groupby("Decade")["GDPGrowth"].mean().sort_index()

avg_growth

Decade
1990    2.645822
2000    4.050309
2010    3.005701
Name: GDPGrowth, dtype: float64

**Interpretation:**

The analysis shows the average GDP growth rate for the countries in the sample across three decades. The 2000s had the highest average growth rate (4.05%), while the 2010s had the lowest (3.01%), suggesting a slowdown in average economic expansion leading up to 2020.

## TASK 4 – Sorting and Extremes

In [15]:
#4.1 	Find the observation with the highest life expectancy and lowest GDP per capita.

#4.1.1 Observation with highest life expectancy
highest_life = df_sampled_data.loc[df_sampled_data["LifeExpectancy"].idxmax()]

#4.1.2 Observation with lowest GDP per capita
lowest_gdp = df_sampled_data.loc[df_sampled_data["GDPperCapita"].idxmin()]

print("Highest Life Expectancy:")
print(highest_life)

print("\nLowest GDP per Capita:")
print(lowest_gdp)

Highest Life Expectancy:
Country                San Marino
Year                         2018
LifeExpectancy             85.095
GDPperCapita         47950.688056
Population                34522.0
SchoolEnrollment         95.12605
InternetUsers                 NaN
HealthExpenditure     3969.364746
LiteracyRate                  NaN
GDPGrowth                1.493905
Decade                       2010
Name: 1151, dtype: object

Lowest GDP per Capita:
Country                 Myanmar
Year                       1992
LifeExpectancy           57.681
GDPperCapita          58.889674
Population           40950342.0
SchoolEnrollment      118.07579
InternetUsers               NaN
HealthExpenditure           NaN
LiteracyRate                NaN
GDPGrowth              9.660942
Decade                     1990
Name: 465, dtype: object


**Interpretation:**

4.1.1 - The highest life expectancy observed in the sample is 85.095 years, recorded in San Marino in the year 2018. This finding highlights San Marino as having one of the world's longest-living populations, a common trait among high-income, small European states that typically exhibit robust healthcare and high living standards.

4.1.2 - Interpretation: The lowest GDP per capita observed in the sample is only $58.89, recorded in Myanmar in the year 1992. This extremely low figure suggests a moment of severe economic underdevelopment or collapse in that country-year, potentially reflecting internal political instability or the difficulties of a newly opening economy.

In [16]:
#4.2 	Display the top 10 records by life expectancy.

# Top 10 records by Life Expectancy
topten_life = df_sampled_data.sort_values("LifeExpectancy", ascending=False).head(10)

topten_life

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade
1151,San Marino,2018,85.095000,47950.688056,34522.0,95.126050,NaN,3969.364746,NaN,1.493905,2010
635,Monaco,2011,85.010000,179363.984222,33946.0,NaN,80.3000,8302.794922,NaN,7.022357,2010
1049,"Hong Kong SAR, China",2017,84.695610,46160.429791,7393200.0,104.807910,89.4159,NaN,NaN,3.796114,2010
688,Andorra,2016,84.489000,40129.819201,72181.0,91.041577,89.7000,2769.152832,NaN,3.709678,2010
542,San Marino,2014,84.394000,51260.493835,32655.0,NaN,NaN,4120.026855,NaN,-0.702371,2010
111,"Hong Kong SAR, China",2016,84.211463,43734.198070,7336600.0,104.741014,87.4794,NaN,NaN,2.175437,2010
33,Japan,2018,84.210976,39751.133098,126811000.0,102.099907,88.7464,4265.779297,NaN,0.643391,2010
1128,San Marino,2013,84.175000,50807.828914,33041.0,NaN,NaN,4113.466797,NaN,-0.783972,2010
1063,Liechtenstein,2019,84.160976,166907.839305,38563.0,103.830009,93.1461,NaN,NaN,NaN,2010
820,"Hong Kong SAR, China",2014,83.935610,40315.373951,7229500.0,105.045690,79.8663,NaN,NaN,2.762430,2010


**Interpretation:** The

top 10 records for life expectancy are consistently found in small, high-income nations like San Marino, Monaco, Andorra, and Liechtenstein, along with developed regions like Japan and Hong Kong SAR, China. The top observation is San Marino in 2018 with a life expectancy of 85.095 years.

## TASK 5 – Counting and Distributions

In [17]:
#5.1 	Count how many observations exist per year and decade.

#5.1.1 Count rows per year
count_per_year = df_sampled_data["Year"].value_counts().sort_index()

print("Observations per Year:")
print(count_per_year)

print("\nObservations per Decade:")

#5.1.2 Create decade column
count_per_de = (df_sampled_data["Year"] // 10) * 10

# Count rows per decade
count_per_decade = count_per_de.value_counts().sort_index()

print(count_per_decade)

Observations per Year:
Year
1990    34
1991    35
1992    38
1993    30
1994    42
1995    52
1996    46
1997    40
1998    40
1999    39
2000    45
2001    39
2002    38
2003    44
2004    42
2005    43
2006    38
2007    34
2008    37
2009    40
2010    46
2011    40
2012    25
2013    39
2014    45
2015    40
2016    46
2017    45
2018    43
2019    38
Name: count, dtype: int64

Observations per Decade:
Year
1990    396
2000    400
2010    407
Name: count, dtype: int64


**Interpretation:** 

5.1.1 - The distribution of observations across the years is relatively even, with most years containing between 30 and 46 records.

5.1.2 - The dataset is highly balanced across the three main decades (1990, 2000, and the 2010s), each containing approximately 400 observations. 

In [18]:
#5.2 	Display the number of unique countries per decade.

# Create decade column
decade_column = (df_sampled_data["Year"] // 10) * 10

# Count unique countries per decade
unique_countries_per_decade = df_sampled_data.groupby(decade_column)["Country"].nunique()

print(unique_countries_per_decade)

Year
1990    212
2000    207
2010    212
Name: Country, dtype: int64


**Interpretation:**

Interpretation:
The dataset covers a consistently large number of unique countries, with 212 countries in both the 1990 and the 2010, and 207 in the 2000. The count of only 29 unique countries for the 2020 is expected since the sample only includes data from the year 2020,


## TASK 6 – Creating New Columns

In [19]:
# Convert population to millions
population = df_sampled_data["Population"] / 1000000

# Compute median and 75th percentile GDP
median_gdp = df_sampled_data["GDPperCapita"].median()
quantile_gdp = df_sampled_data["GDPperCapita"].quantile(0.75)

# Create GDP category
def classify_gdp(gdp):
    if gdp < median_gdp:
        return "Low"
    elif gdp < quantile_gdp:
        return "Medium"
    else:
        return "High"

df_sampled_data["GDPCategory"] = df_sampled_data["GDPperCapita"].apply(classify_gdp)

# Count how many rows fall into each category
category_count = df_sampled_data["GDPCategory"].value_counts()

print("\nCounts per Category:")
print(category_count)



Counts per Category:
GDPCategory
Low       573
High      344
Medium    286
Name: count, dtype: int64


**Interpretation:**

New columns, PopulationMillions and GDPCategory, have been successfully added to the DataFrame. The GDPCategory column buckets countries based on whether their GDP per capita is below the global median (Low), above the 75th percentile (High), or in between (Medium), providing a discrete variable for economic comparisons.

## TASK 7 – Missing Data Analysis

In [20]:
#7.1 	Compute and display missing values (%) for all numeric columns.

# Select numeric columns only
numeric_col = df_sampled_data.select_dtypes(include="number")

# Compute percentage of missing values
missing_percent = numeric_col.isna().mean() * 100

print("Missing Values (%):")
print(missing_percent)

Missing Values (%):
Year                  0.000000
LifeExpectancy        0.000000
GDPperCapita          4.738155
Population            0.000000
SchoolEnrollment     29.093932
InternetUsers        27.015794
HealthExpenditure    40.565254
LiteracyRate         76.142976
GDPGrowth             6.566916
Decade                0.000000
dtype: float64


**Interpretation:**

The results reveal significant data sparsity in several key variables, particularly LiteracyRate (missing nearly 76% of values) and HealthExpenditure (missing nearly 40% of values). In contrast, primary indicators like Year, LifeExpectancy, and Population are complete (0% missing).

In [21]:
#7.2 	Identify which variable has the most missing data.

# Count missing values per column
missing_data = df_sampled_data.isna().sum()

# Identify the column with the most missing values
most_missing = missing_data.idxmax()
count_missing = missing_data.max()

print("Column with the most missing values:", most_missing)
print("Number of missing values:", count_missing)


Column with the most missing values: LiteracyRate
Number of missing values: 916


**Interpretation:**

The LiteracyRate column has the most missing values, indicating that it is likely unavailable for most country-year observations and may be difficult to use in comprehensive analysis.

In [22]:
#7.3 	Count rows with no missing values vs at least one missing.

#7.3.1 Count rows with no missing values
no_missing_val = df_sampled_data.dropna().shape[0]

#7.3.2 Count rows with at least one missing value
at_least_one_missing = df_sampled_data.shape[0] - no_missing_val

print("Rows with NO missing values:", no_missing_val)
print("Rows with AT LEAST one missing value:", at_least_one_missing)

Rows with NO missing values: 158
Rows with AT LEAST one missing value: 1045


**Interpretation:**

7.3.1 - Given that only 158 rows have no missing data, a large portion of the analysis must account for or handle missing values to avoid discarding most of the dataset.

7.3.2 - The resulting value, 1074, signifies that a vast majority of the observations in the dataset have at least one missing value across the 10 available columns.

## TASK 8 – Conditional Selections

In [23]:
#8.1 	Select all rows where LifeExpectancy > 80

# Filter rows where Life Expectancy is greater than 80
high_life = df_sampled_data[df_sampled_data["LifeExpectancy"] > 80]

high_life

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade,GDPCategory
2,United Kingdom,2017,81.256098,40629.234035,65966000.0,100.781052,90.4246,3907.822266,NaN,2.656505,2010,High
33,Japan,2018,84.210976,39751.133098,126811000.0,102.099907,88.7464,4265.779297,NaN,0.643391,2010,High
36,Israel,2010,81.602439,31439.088703,7623600.0,97.658630,67.5000,2237.907959,NaN,5.534340,2010,High
41,San Marino,2008,82.894000,75902.132052,31662.0,94.248047,54.5200,4378.248535,NaN,-0.500939,2000,High
55,Slovenia,2015,80.775610,20697.274853,2063531.0,98.381592,73.0987,1775.860718,NaN,2.393306,2010,High
...,...,...,...,...,...,...,...,...,...,...,...,...
1203,Gibraltar,2013,82.586000,NaN,32407.0,NaN,NaN,NaN,NaN,NaN,2010,High
1204,Israel,2005,80.151220,21286.838864,6930100.0,95.504532,25.1940,1509.105835,NaN,4.387789,2000,High
1205,Monaco,1998,81.605000,90947.205122,32267.0,132.251526,NaN,NaN,NaN,3.503266,1990,High
1212,"Macao SAR, China",2018,83.678049,83761.817180,658900.0,101.238255,83.7941,NaN,NaN,6.405279,2010,High


**Interpretation:**

A total of 109 observations show a life expectancy greater than 80 years. These records predominantly belong to the most developed countries, indicating a strong correlation between life expectancy and economic development.

In [ ]:
#8.2 	Select rows where HealthExpenditure > 8% and GDPGrowth > 3%.

# Select rows where both conditions are true
filtered_rows = df_sampled_data[(df_sampled_data["HealthExpenditure"] > 8) & (df_sampled_data["GDPGrowth"] > 3)]

filtered_rows

**Interpretation:**

A total of 427 observations meet both criteria, representing country-year instances where there was simultaneously a significant investment in health (relative to the threshold) and robust economic expansion.

In [ ]:
#8.3 	Compare the average literacy rate between these two subsets.

# Subset High health expenditure and high GDP growth
subsets = df_sampled_data[(df_sampled_data["HealthExpenditure"] > 8) & (df_sampled_data["GDPGrowth"] > 3)]

# Compute average literacy rate for both subsets
avg_literacy = subsets["LiteracyRate"].mean()

print("Average Literacy Rate:", avg_literacy)

**Interpretation:**

The average literacy rate for these high-growth and high-health-spending observations is approximately 79.62. This finding suggests a positive connection between robust economic activity, health investment, and education, as the countries experiencing these favorable economic conditions also tend to have relatively high rates of literacy.

## TASK 9 – NumPy Array Operations

Import Numpy

In [ ]:
import numpy as np

#9.1 	Compute min, max, mean, and range for LifeExpectancy.

# Convert column to NumPy array
life = df_sampled_data["LifeExpectancy"].to_numpy()

# Compute statistics
min_life = np.min(life)
max_life = np.max(life)
mean_life = np.mean(life)
range_life = max_life - min_life

print("Min Life Expectancy:", min_life)
print("Max Life Expectancy:", max_life)
print("Mean Life Expectancy:", mean_life)
print("Range:", range_life)

**Interpretation:**

The average life expectancy across all observations is approximately 68.79 years, with a wide range of 53.59 years between the minimum (31.51 years) and the maximum (85.09 years). This vast range indicates significant global disparities in health outcomes and development levels within the sample.

In [ ]:
#9.2 	Compute 25th, 50th, and 75th percentiles.

# Compute percentiles
num_25 = np.percentile(life, 25)
num_50 = np.percentile(life, 50)   
num_75 = np.percentile(life, 75)

print("25th Percentile:", num_25)
print("50th Percentile:", num_50)
print("75th Percentile:", num_75)

**Interpretation:**

The median life expectancy (50th percentile) is 70.695 years, meaning half of all country-year observations have a life expectancy below this value. The interquartile range (IQR), between 63.268 years (25th percentile) and 75.752 years (75th percentile), captures the middle 50\% of the data.

In [ ]:
#9.3 	Compute correlation between GDPperCapita and LifeExpectancy.

df_clean = df_sampled_data.dropna(subset=['GDPperCapita', 'LifeExpectancy'])

gdp = df_clean['GDPperCapita'].values
life = df_clean['LifeExpectancy'].values

correlation = np.corrcoef(gdp, life)[0, 1]
print("Correlation between GDPperCapita and LifeExpectancy:", correlation)

**Interpretation:**

The Pearson correlation coefficient of 0.5744 indicates a moderately strong positive linear relationship between a country's GDP per capita and its life expectancy. This suggests that as a country's wealth increases, there is a clear tendency for the average life expectancy of its population to also increase, reflecting the benefits of economic development on health outcomes.

## TASK 10 – Compare 1990 vs 2020

In [3]:
#10.1 	Identify countries with data for both 1990 and 2020.

# Filter data for years 1990 and 2020
df_filtered = df_sampled_data[df_sampled_data['Year'].isin([1990, 2020])]

# Group by 'Country' and count the unique years
year_counts = df_filtered.groupby('Country')['Year'].nunique()

# Identify countries with a count of 2 (meaning they have both 1990 and 2020 data)
countries_both_years_series = year_counts[year_counts == 2].index

# Convert the resulting index (list of countries) to a NumPy array
both_years = countries_both_years_series.to_numpy()

print("Countries with data for both 1990 and 2020 (NumPy Array):")
print(both_years)

Countries with data for both 1990 and 2020 (NumPy Array):
[]


**Interpretation:**

The analysis found no countries with data for both the year 1990 and the year 2020 in the sample, resulting in an empty NumPy array. This indicates that a direct, two-point comparison across these three decades is not possible with this subset of the data.

In [ ]:
#10.2 	Compute change in LifeExpectancy for those countries.

# Since no countries were identified, the change cannot be computed.

if len(both_years) == 0:
    print("Cannot compute change in LifeExpectancy: No countries were identified with data for both 1990 and 2020.")

**Interpretation:**

As confirmed in the previous step, the change in Life Expectancy between 1990 and 2020 cannot be calculated because the sample contains no country that has data points for both specific years.

In [5]:
#10.3 	List the top 3 countries with the greatest improvement.

both_years.head(3)

AttributeError: 'numpy.ndarray' object has no attribute 'head'

**Interpretation:**

The analysis confirms the previous finding: no countries in the sample have a recorded observation for both the exact year 1990 and the exact year 2020.

## Thank you!